pkg/controller/etcdproxy/certs.go

```diff
@@ -7,198 +7,350 @@ import (
+// The Client certificate/key pairs are stored in the Secrets defined in the EtcdStorage Spec.
+// The EtcdProxy controller assumes Secrets for Client certificates are already created, but if not, the controller will try to create them.
+// Creating Secrets for Client certificates requires the appropriate RBAC roles if RBAC is enabled on cluster.
+//
+// This function reads all Secrets provided in the EtcdStorage Spec and checks is the 'etcd.xmudrii.com/certificate-expiry-date'
+// annotation present and contains valid date. If date is in the past, i.e. certificate is expired, or the annotation is not present controller:
+// * Generates new CA certificate. If CA bundle already exists in the controller namespace, the new CA certificate will be appended to the bundle.
+// Expired CA certificates from the bundle are removed in this phase.
+// * Generates new Client certificate/key pair using the newly generated CA certificate and updates the appropriate Secret with new pair.
+// Problem: the API server have to be "restarted" manually to pick up changes. Hopefully, this to be fixed in future Kube versions.
+func (c *EtcdProxyController) ensureClientCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
+	var clientCA *certs.Certificate
+	var errs []error
+	// TODO: Huh, clientCertSecret and secretClientCert sounds way too similar.
+	for _, clientCertSecret := range etcdstorage.Spec.ClientCertSecrets {
+		secret, err := c.getClientCertsSecret(etcdstorage, clientCertSecret)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
 
-	// TODO: implement certificates regeneration. this could require architecture change, because we could require
-	// new secret for storing singer key.
-	serverSigner, err := certs.NewCACertificate(pkix.Name{
-		CommonName: fmt.Sprintf("%s-server-signer-%v", serviceUrl, time.Now().Unix()),
-	}, r.Int63n(100000), currentTime)
+		// Check is annotation containing expiry date present and valid. If it is valid, we're skipping this iteration.
+		if expiry, ok := secret.Annotations[ProxyCertificateExpiryAnnotation]; ok {
+			certExpiry, err := time.Parse(time.RFC3339, expiry)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			if certExpiry.After(time.Now()) {
+				continue
+			}
+		}
+
+		// Generate Client CA certificate and append it to the bundle.
+		// The bundle is located in the ConfigMap in the controller namespace.
+		// We're generating one Client CA in this loop for all Client certificates that are going to be regenerated in this loop/iteration.
+		// TODO: Decide is this a good idea. Another idea is to genearte new CA certificate for each Client certificate that has to be regenrated.
+		// TODO: The negative side in this case is that we could have to restart etcd-proxy for each certificate appended and to get/update for each one instead of once.
+		if clientCA == nil {
+			// Generate new Client CA certificate.
+			clientCA, err = c.generateClientCACertificate(etcdstorage)
```

Comment:
is this a certkeypair?  Maybe call it that?  is this a `generateClientSigningCertKeyPair` method?

Reply:
:+1: